codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/OperationMetricsStruct.java

```diff
@@ -0,0 +1,187 @@
+                        return nil, err
+                    }
+                    om.SignRequestDuration, err = operationMetricTimer(meter, "client.call.auth.signing_duration",
+                        "The time it takes to sign a request")
+                    if err != nil {
+                        return nil, err
+                    }
+                    om.DeserializeDuration, err = operationMetricTimer(meter, "client.call.deserialization_duration",
+                        "The time it takes to deserialize a message body")
+                    if err != nil {
+                        return nil, err
+                    }
+
+                    return context.WithValue(parent, operationMetricsKey{}, om), nil
+                }
+
+                func operationMetricTimer(m metrics.Meter, name, desc string) (metrics.Float64Histogram, error) {
+                    return m.Float64Histogram(name, func(o *metrics.InstrumentOptions) {
+                        o.UnitLabel = "s"
+                        o.Description = desc
+                    })
+                }
+
+                func getOperationMetrics(ctx context.Context) *operationMetrics {
+                    return ctx.Value(operationMetricsKey{}).(*operationMetrics)
+                }
+                """, scope);
+    }
+
+    private GoWriter.Writable generateHelpers() {
+        return goTemplate("""
+                func (m *operationMetrics) histogramFor(name string) metrics.Float64Histogram {
+                    switch name {
+                    case "client.call.duration":
+                        return m.Duration
+                    case "client.call.serialization_duration":
+                        return m.SerializeDuration
+                    case "client.call.resolve_identity_duration":
+                        return m.GetIdentityDuration
+                    case "client.call.resolve_endpoint_duration":
+                        return m.ResolveEndpointDuration
+                    case "client.call.signing_duration":
+                        return m.SignRequestDuration
+                    case "client.call.deserialization_duration":
+                        return m.DeserializeDuration
+                    default:
+                        panic("unrecognized operation metric")
+                    }
+                }
+
+                func timeOperationMetric[T any](
+                    ctx context.Context, metric string, fn func() (T, error),
+                    opts ...metrics.RecordMetricOption,
+                ) (T, error) {
+                    instr := getOperationMetrics(ctx).histogramFor(metric)
+                    opts = append([]metrics.RecordMetricOption{withOperationMetadata(ctx)}, opts...)
+
+                    start := time.Now()
+                    v, err := fn()
+                    end := time.Now()
+
+                    elapsed := end.Sub(start)
+                    instr.Record(ctx, float64(elapsed)/1e9, opts...)
```

Comment:
does the SRA mention what unit should the duration have? Seconds seems a bit too coarse, and ideally I'd like us to just pass a `time.Duration` object without doing math on top of it

Reply:
Discussed offline, this is an OTEL spec mandate.

Reply:
But for the record I completely agree. It should just be nanoseconds (most atomic unit) and observability backends should figure it out(tm)